Updated LOG/TLM buffer size limit

bevinduckett opened the pull request:
| | |
|:---|:---|
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  n|
|**_Documentation Included (y/n)_**|  n|
|**_Generative AI was used in this contribution (y/n)_**|  n|

---
## Change Description

1. Added a `FW_TIME_SERIALIZED_SIZE`  macro to FpConfig.h
2. Updated the calculation for `FW_LOG_BUFFER_MAX_SIZE` & `FW_TLM_BUFFER_MAX_SIZE`  in `FpConfig.h` to use the new macro 
3. Added an assert to Time.cpp to verify the macro's value was not out of date (if macro was defined) 
4. Updated the static assert in `FPrimeBasicTypes.hpp` that checks `FW_TLM_STRING_MAX_SIZE` to use `FwTlmPacketizeIdType` instead of `FwSizeStoreType` 

## Rationale
According to `docs/user-manual/framework/configuring-fprime.md`, `FW_LOG_BUFFER_MAX_SIZE` & `FW_TLM_BUFFER_MAX_SIZE` should be the number of bytes available for argument data in each packet types.

However, the current method of calculating those values does not account for the space used by storing the serialized value of a Fw::Time object into the com buffer. As a result, the static asserts in FPrimeBasicTypes.hpp which check that FW_LOG_STRING_MAX_SIZE & FW_TLM_STRING_MAX_SIZE are sized so they can fit into a com buffer aren't working as intended.
```
static_assert(FW_LOG_STRING_MAX_SIZE + sizeof(FwSizeStoreType) <= FW_LOG_BUFFER_MAX_SIZE,
              "FW_LOG_STRING_MAX_SIZE cannot be larger than FW_LOG_BUFFER_MAX_SIZE");
static_assert(FW_TLM_STRING_MAX_SIZE + sizeof(FwSizeStoreType) <= FW_TLM_BUFFER_MAX_SIZE,
              "FW_TLM_STRING_MAX_SIZE cannot be larger than FW_TLM_BUFFER_MAX_SIZE");
```

The solution I recommend is adding a macro (TIME_SERIALIZED_SIZE), using that macro to calculating FW_TLM_BUFFER_MAX_SIZE/FW_LOG_BUFFER_MAX_SIZE, and then checking that the value of TIME_SERIALIZED_SIZE is correct with a static assert in Time.cpp. (A macro is used instead of relying on Fw::Time::SERIALIZED_SIZE because Time.hpp cannot be included into `FpConfig.h`

While investigating the static asserts in FPrimeBasicTypes.hpp , I also noticed that the static assert for FW_TLM_STRING_MAX_SIZE  was using `sizeof(FwSizeStoreType)` instead of `sizeof(FwTlmPacketizeIdType)` (they're both U16 values so it's functionally correct but I recommend updating to improve clarity)

## Testing/Review Recommendations

To verify the static asserts were not working as expected, set `FW_COM_BUFFER_MAX_SIZE` to 70 and set `TLM_LOG_STRING_MAX_SIZE` to 61, you'll observe that the static assert fails with the below message. Then set `TLM_LOG_STRING_MAX_SIZE ` to 60 and the build will pass.
```
In file included from <path>/fprime/Fw/Types/Assert.hpp:4:
<path>/fprime/Fw/FPrimeBasicTypes.hpp:51:15: error: static assertion failed due to requirement '61 + sizeof(unsigned short) <= (70 - sizeof(unsigned int) - sizeof(unsigned int))': FW_LOG_STRING_MAX_SIZE cannot be larger than FW_LOG_BUFFER_MAX_SIZE
   51 | static_assert(FW_LOG_STRING_MAX_SIZE + sizeof(FwSizeStoreType) <= FW_LOG_BUFFER_MAX_SIZE,
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<build_path>/F-Prime/default/config/../config/FpConfig.h:207:32: note: expanded from macro 'FW_LOG_STRING_MAX_SIZE'
  207 | #define FW_LOG_STRING_MAX_SIZE 61  //!< Max size of log string parameter type
      |                                ^
 <path>/fprime/Fw/FPrimeBasicTypes.hpp:51:64: note: expression evaluates to '63 <= 62'
   51 | static_assert(FW_LOG_STRING_MAX_SIZE + sizeof(FwSizeStoreType) <= FW_LOG_BUFFER_MAX_SIZE,
```
Then compare that to the  `Fw::LogPacket::serializeTo()` function (in Fw/Log/LogPacket.cpp) which stores the com packet base, event/log ID, and then time tag before adding the argument data. The serialized size of time is 11 bytes, so there's no way a 60 byte string could fit into ComBuffer. 

## Future Work

<!-- Note any additional work that will be done relating to this issue. -->

## AI Usage (see [policy](https://github.com/nasa/fprime/blob/devel/AI_POLICY.md))

<!-- If AI was used, please describe how it was utilized (e.g., code generation, documentation, testing, debugging assistance, etc.). -->

bevinduckett committed:
Updated LOG/TLM buffer size limit

According to `docs/user-manual/framework/configuring-fprime.md`, `FW_LOG_BUFFER_MAX_SIZE` & `FW_TLM_BUFFER_MAX_SIZE` should be the number of bytes available for argument data in each packet types.

However, the current method of calculating those values does not account for the space used by storing Fw::Time as a serialized object. As a result, the static asserts in FPrimeBasicTypes.hpp which check that the STRING_MAX_SIZE values for LOG & TLM are correct aren't working as intended.

The solution I recommend is adding a macro (TIME_SERIALIZED_SIZE), using that macro to calculating FW_TLM_BUFFER_MAX_SIZE/FW_LOG_BUFFER_MAX_SIZE, and then checking that the value of TIME_SERIALIZED_SIZE is correct with a static assert in Time.cpp

While investigating the static asserts in FPrimeBasicTypes.hpp , I also noticed that the static assert for FW_TLM_STRING_MAX_SIZE  was using `sizeof(FwSizeStoreType)` instead of `sizeof(FwTlmPacketizeIdType)` (they're both U16 values so it's functionally correct but I recommend updating to improve clarity)

thomas-bc commented on Fw/Time/Time.cpp:
```diff
@@ -2,6 +2,10 @@
 
 namespace Fw {
+#ifdef FW_TIME_SERIALIZED_SIZE
+static_assert(FW_TIME_SERIALIZED_SIZE == Fw::Time::SERIALIZED_SIZE, "FW_TIME_SERIALIZED_SIZE macro is wrong");
```
Might also want to add a static assert for 

```
Fw::Time::SERIALIZED_SIZE == Fw::TimeValue::SERIALIZED_SIZE
```

which is the FPP type ([Fw.TimeValue](https://github.com/nasa/fprime/blob/b377be2a9fda0cfcb3896c6ed3cdfe22c0af27ff/Fw/Time/Time.fpp#L4-L9))

bevinduckett replied:
@thomas-bc  - makes sense, I've added that assert in https://github.com/nasa/fprime/pull/4179/commits/f5d21551256c46f41713a43450ca23213651d65a

bevinduckett committed:
Fixed formatting of Time.cpp & FpConfig.h

bevinduckett commented on Fw/Time/Time.cpp:
```diff
@@ -2,6 +2,10 @@
 #include <Fw/Time/Time.hpp>
 
 namespace Fw {
+#ifdef FW_TIME_SERIALIZED_SIZE
```
agree in theory - in this case I recommend using the `ifdef` to avoid compatibility issues if a progress is using a FpConfig.h that doesn't have FW_TIME_SERIALIZED_SIZE

celskeggs replied:
We'll probably want to phase out this `#ifdef` eventually. @thomas-bc Do you have a preferred way to phase out workarounds like this once you expect downstream projects to have updated their `FpConfig.h`?

LeStarch replied:
For phasing things like this we are trying  to follow this approach:

1. First release after change: no warning, but ship a release note
2. Second release, add a warning. "Deprecated". 
3. Remove the Deprecated feature. 

However,  for the next release we already have some must-make configuration additions as there was no good solution should a field not be defined.  Thus users will need new fields in configuration anyway.

bevinduckett committed:
Added another static assert to Time.cpp

Added a static assert that the SERIALIZED_SIZE value was the same for Fw::Time and Fw::TimeValue

bevinduckett committed:
Merge branch 'devel' into bduckett-max_buffer_size_fix

LeStarch reviewed (dismissed):
Ideally we'd use the FPP sizes here rather than defining a new variable, which is required to have a matching size.

However, to do this well we'd need to ensure that Fw.Time (which could depend on FpConfig) and FpConfig (which would have to depend on Fw.Time) were in the same build module to avoid circular dependencies. These are currently two separate packages.  

Practically speaking,  this would like require a separate config module for tlm, cmd, and event constants.  Therefore,  I conclude that this is the least ugly solution...but we should track the other work as an issue so we don't forget about it.

bevinduckett committed:
Removed space to fix a formatting error in FpConfig.h

bevinduckett committed:
Merge branch 'devel' into bduckett-max_buffer_size_fix

thomas-bc committed:
fix static assert

bevinduckett commented:
@LeStarch - I updated a file to fix a formatting error, can you please trigger the other CI checks

bevinduckett commented:
@LeStarch - is there something I should do for the CodeQL alert (https://github.com/nasa/fprime/pull/4179/checks?check_run_id=51186999052) re: use of conditional compilation?

bevinduckett commented:
I would also appreciate guidance on how/whether I should address the other CI failures (see below) and/or the spelling issues noted by https://github.com/nasa/fprime/actions/runs/17994241319 (none of which are related to files I changed) 

<img width="522" height="196" alt="image" src="https://github.com/user-attachments/assets/ea528799-4404-432e-addd-3e3043041212" />
<img width="531" height="245" alt="image" src="https://github.com/user-attachments/assets/9f9470fe-6a67-4128-b260-b1a3246d5d61" />
<img width="522" height="195" alt="image" src="https://github.com/user-attachments/assets/270ee148-55f1-4106-8d9d-abad6234e67c" />
<img width="533" height="114" alt="image" src="https://github.com/user-attachments/assets/c66f262e-da51-44c3-8ed2-3b6f54f106bd" />
<img width="519" height="190" alt="image" src="https://github.com/user-attachments/assets/d3bb7f7f-1fdc-45af-86a9-e168efe5ae4d" />

thomas-bc commented:
@bevinduckett it looks like CI is passing. Our internal runners are having some intermittent issues at the moment, hence the (unrelated) failure on that RPI one (for now, trying to get it to pass)

bevinduckett commented:
@thomas-bc , thank you for clarifying. In that case, can this be approved & merged?

LeStarch commented:
I think this work is going to be need to be redone by PR #4194 as all these constants are being moved into the model for FPP.   However, this fix will be needed there.

@bevinduckett my thoughts would be to close here, and open up a ticket to fix in FPP once #4194  and nasa/fpp#725 are done.  Thoughts?

bevinduckett commented:
@LeStarch - that sounds good to me, would you like me to create the new ticket or would you rather handle that? I am about to leave for a three week vacation

LeStarch commented:
@bevinduckett I'll do the issue.
